project3/participant/input_parser/input_parser.cpp

```diff
@@ -0,0 +1,65 @@
+#include "input_parser.h"
+
+namespace participant::input_parser {
+
+InputParser::InputParser()
+    : multi_id_{0}, is_valid_{true}, req_{REG}, message_{} {}
+
+void InputParser::Parse(std::string& cmd) {
+  std::istringstream iss(cmd);
+  std::string word;
+  iss >> word;
+  auto itr = commands_.find(word);
+  is_valid_ = itr != commands_.end();
+  if (!is_valid_) {
+    return;
+  }
+  req_ = itr->second;
+  switch (req_) {
+    case MSEND:
+      std::getline(iss, message_);
+      break;
+    case REG:
+    case RECON:
+    case DISCON:
+    case DEREG:
+      break;
+  }
+}  // InputParser
+
+bool InputParser::IsValid() { return is_valid_; }
+
+google::protobuf::Message* InputParser::CreateReq() {
+  switch (req_) {
+    case REG:
+      return CreateRegMsg();
+    case DEREG:
+      return CreateDeregMsg();
+    case DISCON:
+      return CreateDisconMsg();
+    case RECON:
+      return CreateReconMsg();
+    case MSEND:
+      return CreateMsendMsg();
+  }
+  return nullptr;
+}
+pub_sub_messages::Register* InputParser::CreateRegMsg() {
+  return pub_sub_messages::Register().New();
```

Comment:
The Protobuf documentation is a little vague on this, but I think [messages created with `New()` have to be manually deleted](https://stackoverflow.com/questions/10217436/properly-freeing-protocol-buffer-memory/10217744). How about just creating an instance of each message type as class members and handing out pointers to those?